File: src/modules/marketing/masterData.ts
/**
 * Institution Master Data — the ten groups from the Berana LMS Institution
 * Master Data Collection Guide, mirroring the backend schema in
 * `app/modules/onboarding/master_data.py`.
 *
 * Institution ID is deliberately absent: it is generated server-side ("INST-0001")
 * and a value sent from here is ignored.
 */

export type OwnershipType =
  | 'public'
  | 'private'
  | 'government'
  | 'ngo'
  | 'religious'
  | 'other'

export type TermStructure = 'semester' | 'trimester' | 'quarter' | 'year' | 'continuous'

export type BillingCycle = 'annual' | 'semi_annual' | 'quarterly' | 'monthly'

export interface MasterDataForm {
  identity: {
    institution_code: string
    short_name: string
    ownership_type: OwnershipType | ''
    registration_number: string
    year_established: string
  }
  contact: {
    main_phone: string
    secondary_phone: string
    official_email: string
    website: string
    inquiry_email: string
    admissions_email: string
    registrar_email: string
  }
  address: {
    country: string
    state_region: string
    city: string
    sub_city: string
    street: string
    postal_code: string
  }
  primary_contact: {
    full_name: string
    job_title: string
    department: string
    email: string
    phone: string
  }
  leadership: {
    president: string
    vice_president: string
    academic_head: string
    registrar: string
    hr_contact: string
    finance_contact: string
  }
  academic_structure: {
    campus_count: string
    campuses: string
    colleges: string
    faculties: string
    departments: string
    academic_units: string
  }
  academic_calendar: {
    current_academic_year: string
    start_date: string
    end_date: string
    term_structure: TermStructure | ''
    term_count: string
    registration_start: string
    registration_end: string
    class_start: string
    class_end: string
    exam_start: string
    exam_end: string
  }
  certification: {
    certificate_types: string
    signatory_name: string
    signatory_title: string
    numbering_format: string
    has_institutional_seal: boolean
  }
  lms_administrator: {
    full_name: string
    job_title: string
    department: string
    email: string
    phone: string
    administrator_role: string
  }
  subscription: {
    package: string
    license_type: string
    contracted_users: string
    billing_cycle: BillingCycle | ''
  }
}

export const EMPTY_MASTER_DATA: MasterDataForm = {
  identity: {
    institution_code: '',
    short_name: '',
    ownership_type: '',
    registration_number: '',
    year_established: '',
  },
  contact: {
    main_phone: '',
    secondary_phone: '',
    official_email: '',
    website: '',
    inquiry_email: '',
    admissions_email: '',
    registrar_email: '',
  },
  address: {
    country: '',
    state_region: '',
    city: '',
    sub_city: '',
    street: '',
    postal_code: '',
  },
  primary_contact: { full_name: '', job_title: '', department: '', email: '', phone: '' },
  leadership: {
    president: '',
    vice_president: '',
    academic_head: '',
    registrar: '',
    hr_contact: '',
    finance_contact: '',
  },
  academic_structure: {
    campus_count: '',
    campuses: '',
    colleges: '',
    faculties: '',
    departments: '',
    academic_units: '',
  },
  academic_calendar: {
    current_academic_year: '',
    start_date: '',
    end_date: '',
    term_structure: '',
    term_count: '',
    registration_start: '',
    registration_end: '',
    class_start: '',
    class_end: '',
    exam_start: '',
    exam_end: '',
  },
  certification: {
    certificate_types: '',
    signatory_name: '',
    signatory_title: '',
    numbering_format: '',
    has_institutional_seal: false,
  },
  lms_administrator: {
    full_name: '',
    job_title: '',
    department: '',
    email: '',
    phone: '',
    administrator_role: 'Institution Admin',
  },
  subscription: { package: '', license_type: '', contracted_users: '', billing_cycle: '' },
}

export const OWNERSHIP_OPTIONS: { value: OwnershipType; label: string }[] = [
  { value: 'public', label: 'Public' },
  { value: 'private', label: 'Private' },
  { value: 'government', label: 'Government' },
  { value: 'ngo', label: 'NGO / Non-profit' },
  { value: 'religious', label: 'Religious' },
  { value: 'other', label: 'Other' },
]

export const TERM_STRUCTURE_OPTIONS: { value: TermStructure; label: string }[] = [
  { value: 'semester', label: 'Semester' },
  { value: 'trimester', label: 'Trimester' },
  { value: 'quarter', label: 'Quarter' },
  { value: 'year', label: 'Full year' },
  { value: 'continuous', label: 'Continuous / rolling' },
]

export const BILLING_CYCLE_OPTIONS: { value: BillingCycle; label: string }[] = [
  { value: 'annual', label: 'Annual' },
  { value: 'semi_annual', label: 'Semi-annual' },
  { value: 'quarterly', label: 'Quarterly' },
  { value: 'monthly', label: 'Monthly' },
]

/** Labels for the console's read-only view, in the guide's order. */
export const MASTER_DATA_SECTIONS: { key: keyof MasterDataForm; label: string }[] = [
  { key: 'identity', label: 'Institution Identity' },
  { key: 'contact', label: 'Contact Information' },
  { key: 'address', label: 'Physical Address' },
  { key: 'primary_contact', label: 'Primary Contact Person' },
  { key: 'leadership', label: 'Institution Leadership' },
  { key: 'academic_structure', label: 'Academic Structure' },
  { key: 'academic_calendar', label: 'Academic Calendar' },
  { key: 'certification', label: 'Certification Information' },
  { key: 'lms_administrator', label: 'Berana LMS Administrator' },
  { key: 'subscription', label: 'Subscription & Licensing' },
]

export const MASTER_DATA_FIELD_LABELS: Record<string, string> = {
  institution_code: 'Institution code',
  short_name: 'Short name / acronym',
  ownership_type: 'Ownership type',
  registration_number: 'Registration / license number',
  year_established: 'Year established',
  main_phone: 'Main telephone',
  secondary_phone: 'Secondary telephone',
  official_email: 'Official email',
  website: 'Website',
  inquiry_email: 'General inquiry email',
  admissions_email: 'Admissions email',
  registrar_email: 'Registrar email',
  country: 'Country',
  state_region: 'State / region',
  city: 'City',
  sub_city: 'Sub-city / district',
  street: 'Street address',
  postal_code: 'Postal / ZIP code',
  full_name: 'Full name',
  job_title: 'Job title',
  department: 'Department / office',
  email: 'Email',
  phone: 'Telephone',
  president: 'President / CEO / Director',
  vice_president: 'Vice president / deputy',
  academic_head: 'Academic head',
  registrar: 'Registrar',
  hr_contact: 'HR contact',
  finance_contact: 'Finance contact',
  campus_count: 'Number of campuses',
  campuses: 'Campuses',
  colleges: 'Colleges / schools',
  faculties: 'Faculties',
  departments: 'Departments',
  academic_units: 'Academic units',
  current_academic_year: 'Current academic year',
  start_date: 'Academic year start',
  end_date: 'Academic year end',
  term_structure: 'Term structure',
  term_count: 'Number of terms',
  registration_start: 'Registration opens',
  registration_end: 'Registration closes',
  class_start: 'Classes start',
  class_end: 'Classes end',
  exam_start: 'Examinations start',
  exam_end: 'Examinations end',
  certificate_types: 'Certificate types',
  signatory_name: 'Authorized signatory',
  signatory_title: 'Signatory title',
  numbering_format: 'Certificate numbering format',
  has_institutional_seal: 'Institutional seal available',
  administrator_role: 'Administrator role',
  package: 'Berana package',
  license_type: 'License type',
  contracted_users: 'Contracted users',
  billing_cycle: 'Billing cycle',
}

const csv = (value: string): string[] =>
  value
    .split(',')
    .map((part) => part.trim())
    .filter(Boolean)

const text = (value: string): string | undefined => value.trim() || undefined

const num = (value: string): number | undefined => {
  const parsed = Number(value.trim())
  return value.trim() && Number.isFinite(parsed) ? parsed : undefined
}

/**
 * Convert the form (all strings, as HTML inputs produce) into the JSON the API
 * expects. Empty fields are dropped rather than sent as "", which the backend
 * would reject for typed fields such as dates and enums.
 */
export function toMasterDataPayload(form: MasterDataForm): Record<string, unknown> {
  return {
    identity: {
      institution_code: text(form.identity.institution_code),
      short_name: text(form.identity.short_name),
      ownership_type: form.identity.ownership_type || undefined,
      registration_number: text(form.identity.registration_number),
      year_established: num(form.identity.year_established),
    },
    contact: {
      main_phone: text(form.contact.main_phone),
      secondary_phone: text(form.contact.secondary_phone),
      official_email: text(form.contact.official_email),
      website: text(form.contact.website),
      inquiry_email: text(form.contact.inquiry_email),
      admissions_email: text(form.contact.admissions_email),
      registrar_email: text(form.contact.registrar_email),
    },
    address: {
      country: text(form.address.country),
      state_region: text(form.address.state_region),
      city: text(form.address.city),
      sub_city: text(form.address.sub_city),
      street: text(form.address.street),
      postal_code: text(form.address.postal_code),
    },
    primary_contact: {
      full_name: text(form.primary_contact.full_name),
      job_title: text(form.primary_contact.job_title),
      department: text(form.primary_contact.department),
      email: text(form.primary_contact.email),
      phone: text(form.primary_contact.phone),
    },
    leadership: {
      president: text(form.leadership.president),
      vice_president: text(form.leadership.vice_president),
      academic_head: text(form.leadership.academic_head),
      registrar: text(form.leadership.registrar),
      hr_contact: text(form.leadership.hr_contact),
      finance_contact: text(form.leadership.finance_contact),
    },
    academic_structure: {
      campus_count: num(form.academic_structure.campus_count),
      campuses: csv(form.academic_structure.campuses),
      colleges: csv(form.academic_structure.colleges),
      faculties: csv(form.academic_structure.faculties),
      departments: csv(form.academic_structure.departments),
      academic_units: csv(form.academic_structure.academic_units),
    },
    academic_calendar: {
      current_academic_year: text(form.academic_calendar.current_academic_year),
      start_date: text(form.academic_calendar.start_date),
      end_date: text(form.academic_calendar.end_date),
      term_structure: form.academic_calendar.term_structure || undefined,
      term_count: num(form.academic_calendar.term_count),
      registration_start: text(form.academic_calendar.registration_start),
      registration_end: text(form.academic_calendar.registration_end),
      class_start: text(form.academic_calendar.class_start),
      class_end: text(form.academic_calendar.class_end),
      exam_start: text(form.academic_calendar.exam_start),
      exam_end: text(form.academic_calendar.exam_end),
    },
    certification: {
      certificate_types: csv(form.certification.certificate_types),
      signatory_name: text(form.certification.signatory_name),
      signatory_title: text(form.certification.signatory_title),
      numbering_format: text(form.certification.numbering_format),
      has_institutional_seal: form.certification.has_institutional_seal,
    },
    lms_administrator: {
      full_name: form.lms_administrator.full_name.trim(),
      job_title: text(form.lms_administrator.job_title),
      department: text(form.lms_administrator.department),
      email: form.lms_administrator.email.trim(),
      phone: text(form.lms_administrator.phone),
      administrator_role: form.lms_administrator.administrator_role.trim() || 'Institution Admin',
    },
    subscription: {
      package: text(form.subscription.package),
      license_type: text(form.subscription.license_type),
      contracted_users: num(form.subscription.contracted_users),
      billing_cycle: form.subscription.billing_cycle || undefined,
    },
  }
}

/**
 * The guide's required minimum, checked in the browser so the user is told which
 * step to go back to. The backend enforces the same rules regardless.
 */
export function missingRequiredFields(form: MasterDataForm, institutionName: string): string[] {
  const missing: string[] = []
  if (!institutionName.trim()) missing.push('Legal institution name')
  if (!form.address.country.trim()) missing.push('Country')
  if (!form.contact.official_email.trim()) missing.push('Official institution email')
  if (!form.primary_contact.full_name.trim()) missing.push('Primary contact name')
  if (!form.primary_contact.email.trim()) missing.push('Primary contact email')
  if (!form.lms_administrator.full_name.trim()) missing.push('LMS administrator name')
  if (!form.lms_administrator.email.trim()) missing.push('LMS administrator email')
  return missing
}

File: backend/app/modules/onboarding/master_data.py
"""
Institution Master Data (Master_Data.pdf, September 2026).

The ten data groups an institution fills in when it registers. They are stored as
one validated JSONB document on the service request rather than forty columns:
the shape is fixed here by Pydantic, so it is still typed and validated, but the
guide can gain a field without a migration.

Per the guide's development notes:

* ``institution_ref`` ("INST-0001") is system-generated and unique — never taken
  from the form.
* The minimum required set is legal name, institution type, country, official
  email, primary contact and LMS administrator. Everything else may be filled in
  later by the institution admin.
* Branding, programs, courses, grading and delivery model are deliberately out of
  scope; they are configured inside the workspace after activation.
"""
from __future__ import annotations

from datetime import date
from enum import Enum

from pydantic import BaseModel, ConfigDict, EmailStr, Field, field_validator


class OwnershipType(str, Enum):
    PUBLIC = "public"
    PRIVATE = "private"
    GOVERNMENT = "government"
    NGO = "ngo"
    RELIGIOUS = "religious"
    OTHER = "other"


class TermStructure(str, Enum):
    SEMESTER = "semester"
    TRIMESTER = "trimester"
    QUARTER = "quarter"
    YEAR = "year"
    CONTINUOUS = "continuous"


class BillingCycle(str, Enum):
    ANNUAL = "annual"
    SEMI_ANNUAL = "semi_annual"
    QUARTERLY = "quarterly"
    MONTHLY = "monthly"


class _Section(BaseModel):
    """Sections are optional as a whole; only the guide's minimum set is required."""

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


# 1. Institution Identity — legal name and institution type live on the service
#    request itself, so they are not repeated here.
class InstitutionIdentity(_Section):
    institution_code: str | None = Field(default=None, max_length=32)
    short_name: str | None = Field(default=None, max_length=80)
    ownership_type: OwnershipType | None = None
    registration_number: str | None = Field(default=None, max_length=80)
    year_established: int | None = Field(default=None, ge=1000, le=2200)


# 2. Contact Information
class ContactInformation(_Section):
    main_phone: str | None = Field(default=None, max_length=50)
    secondary_phone: str | None = Field(default=None, max_length=50)
    official_email: EmailStr | None = None
    website: str | None = Field(default=None, max_length=200)
    inquiry_email: EmailStr | None = None
    admissions_email: EmailStr | None = None
    registrar_email: EmailStr | None = None


# 3. Physical Address
class PhysicalAddress(_Section):
    country: str | None = Field(default=None, max_length=100)
    state_region: str | None = Field(default=None, max_length=100)
    city: str | None = Field(default=None, max_length=100)
    sub_city: str | None = Field(default=None, max_length=100)
    street: str | None = Field(default=None, max_length=250)
    postal_code: str | None = Field(default=None, max_length=30)


# 4. Primary Contact Person
class PrimaryContact(_Section):
    full_name: str | None = Field(default=None, max_length=200)
    job_title: str | None = Field(default=None, max_length=150)
    department: str | None = Field(default=None, max_length=150)
    email: EmailStr | None = None
    phone: str | None = Field(default=None, max_length=50)


# 5. Institution Leadership
class Leadership(_Section):
    president: str | None = Field(default=None, max_length=200)
    vice_president: str | None = Field(default=None, max_length=200)
    academic_head: str | None = Field(default=None, max_length=200)
    registrar: str | None = Field(default=None, max_length=200)
    hr_contact: str | None = Field(default=None, max_length=200)
    finance_contact: str | None = Field(default=None, max_length=200)


# 6. Academic Structure
class AcademicStructure(_Section):
    campus_count: int | None = Field(default=None, ge=0, le=1000)
    campuses: list[str] = Field(default_factory=list, max_length=200)
    colleges: list[str] = Field(default_factory=list, max_length=200)
    faculties: list[str] = Field(default_factory=list, max_length=200)
    departments: list[str] = Field(default_factory=list, max_length=500)
    academic_units: list[str] = Field(default_factory=list, max_length=500)


# 7. Academic Calendar
class AcademicCalendar(_Section):
    current_academic_year: str | None = Field(default=None, max_length=20)
    start_date: date | None = None
    end_date: date | None = None
    term_structure: TermStructure | None = None
    term_count: int | None = Field(default=None, ge=1, le=12)
    registration_start: date | None = None
    registration_end: date | None = None
    class_start: date | None = None
    class_end: date | None = None
    exam_start: date | None = None
    exam_end: date | None = None

    @field_validator("end_date", "registration_end", "class_end", "exam_end")
    @classmethod
    def _not_before_start(cls, value: date | None, info):
        starts = {
            "end_date": "start_date",
            "registration_end": "registration_start",
            "class_end": "class_start",
            "exam_end": "exam_start",
        }
        start = info.data.get(starts[info.field_name])
        if value and start and value < start:
            raise ValueError(f"{info.field_name} cannot be before {starts[info.field_name]}")
        return value


# 8. Certification Information
class CertificationInfo(_Section):
    certificate_types: list[str] = Field(default_factory=list, max_length=50)
    signatory_name: str | None = Field(default=None, max_length=200)
    signatory_title: str | None = Field(default=None, max_length=150)
    numbering_format: str | None = Field(default=None, max_length=100)
    has_institutional_seal: bool | None = None


# 9. Berana LMS Administrator — becomes the institution admin account at activation.
class LmsAdministrator(_Section):
    full_name: str = Field(min_length=1, max_length=200)
    job_title: str | None = Field(default=None, max_length=150)
    department: str | None = Field(default=None, max_length=150)
    email: EmailStr
    phone: str | None = Field(default=None, max_length=50)
    administrator_role: str = Field(default="Institution Admin", max_length=80)


# 10. Subscription & Licensing — the institution states what it wants; activation
#     dates and renewal are set by the console when Cyber-Zeb activates.
class SubscriptionRequest(_Section):
    package: str | None = Field(default=None, max_length=100)
    license_type: str | None = Field(default=None, max_length=100)
    contracted_users: int | None = Field(default=None, ge=1, le=10_000_000)
    billing_cycle: BillingCycle | None = None


class InstitutionMasterData(BaseModel):
    """The full master data document stored on a service request."""

    model_config = ConfigDict(extra="forbid")

    identity: InstitutionIdentity = Field(default_factory=InstitutionIdentity)
    contact: ContactInformation = Field(default_factory=ContactInformation)
    address: PhysicalAddress = Field(default_factory=PhysicalAddress)
    primary_contact: PrimaryContact = Field(default_factory=PrimaryContact)
    leadership: Leadership = Field(default_factory=Leadership)
    academic_structure: AcademicStructure = Field(default_factory=AcademicStructure)
    academic_calendar: AcademicCalendar = Field(default_factory=AcademicCalendar)
    certification: CertificationInfo = Field(default_factory=CertificationInfo)
    lms_administrator: LmsAdministrator
    subscription: SubscriptionRequest = Field(default_factory=SubscriptionRequest)

    @field_validator("address")
    @classmethod
    def _country_required(cls, value: PhysicalAddress) -> PhysicalAddress:
        # Guide, Development Notes: country is part of the required minimum.
        if not (value.country or "").strip():
            raise ValueError("Country is required")
        return value

    @field_validator("contact")
    @classmethod
    def _official_email_required(cls, value: ContactInformation) -> ContactInformation:
        if not value.official_email:
            raise ValueError("Official institution email is required")
        return value

    @field_validator("primary_contact")
    @classmethod
    def _primary_contact_required(cls, value: PrimaryContact) -> PrimaryContact:
        if not (value.full_name or "").strip() or not value.email:
            raise ValueError("Primary contact name and email are required")
        return value


# Display order and labels for the console's read-only view of a submitted record.
SECTION_LABELS: dict[str, str] = {
    "identity": "Institution Identity",
    "contact": "Contact Information",
    "address": "Physical Address",
    "primary_contact": "Primary Contact Person",
    "leadership": "Institution Leadership",
    "academic_structure": "Academic Structure",
    "academic_calendar": "Academic Calendar",
    "certification": "Certification Information",
    "lms_administrator": "Berana LMS Administrator",
    "subscription": "Subscription & Licensing",
}
